File: src/decorator.py
"""
Decorator Pattern
=================

The decorator pattern (not to confused with Python's decorators) is
straight-forward in Python.
"""

from abc import ABC, abstractmethod
import math


class Predictor(ABC):
    @abstractmethod
    def fit(self, x, y): pass

    @abstractmethod
    def predict(self, x): pass


class PredictorDecorator(Predictor):
    def __init__(self, decorated_predictor: Predictor):
        self.decorated_predictor = decorated_predictor

    @abstractmethod
    def fit(self, x, y): pass

    @abstractmethod
    def predict(self, x): pass


class LinearPredictor(Predictor):
    def __init__(self):
        self.slope = 0
        self.offset = 0

    def fit(self, x, y):
        mean_x = sum(x) / len(x)
        mean_y = sum(y) / len(y)
        sos_xy = sum((xi - mean_x) * (yi - mean_y) for xi, yi in zip(x, y))
        sos_x = sum((xi - mean_x) ** 2 for xi in x)
        self.slope = sos_xy / sos_x
        self.offset = mean_y - self.slope * mean_x
        return self

    def predict(self, x):
        return [self.offset + self.slope * xi for xi in x]


class LogYDecorator(PredictorDecorator):
    def fit(self, x, y):
        logy = [math.log(yi) for yi in y]
        self.decorated_predictor.fit(x, logy)
        return self

    def predict(self, x):
        logy = self.decorated_predictor.predict(x)
        return [math.exp(yi) for yi in logy]


class LogXDecorator(PredictorDecorator):
    def fit(self, x, y):
        logx = [math.log(xi) for xi in x]
        self.decorated_predictor.fit(logx, y)
        return self

    def predict(self, x):
        logx = [math.log(xi) for xi in x]
        return self.decorated_predictor.predict(logx)


if __name__ == '__main__':
    x_train = [2, 4, 6, 8]
    y_train = [1, 2, 3, 4]

    linear_predictor = LinearPredictor().fit(x_train, y_train)
    print('linear predictor:', linear_predictor.predict(x_train))

    logy_predictor = LogYDecorator(LinearPredictor()).fit(x_train, y_train)
    print('log-y predictor:', logy_predictor.predict(x_train))

    logx_predictor = LogXDecorator(LinearPredictor()).fit(x_train, y_train)
    print('log-x predictor:', logx_predictor.predict(x_train))

    loglog_predictor = LogXDecorator(LogYDecorator(LinearPredictor())).fit(x_train, y_train)
    print('log-log predictor:', loglog_predictor.predict(x_train))

File: src/decorator.rs
//! Decorator Pattern
//! =================
//!
//! Traits only define behavior and no state, unlike classes in other
//! languages. Thus, it does not make sense to define a separate
//! Decorator interface, that only differs from the Component
//! interface by containing state in the form of the wrapped object.
//! Concrete decorators each need to implement their own state and
//! they can directly use the Component trait.
//!

trait Predictor {
    fn fit(self, x: impl Iterator<Item = f64>, y: impl Iterator<Item = f64>) -> Self;
    fn predict<'a>(&self, x: impl Iterator<Item = f64> + 'a) -> Box<dyn Iterator<Item = f64> + 'a>;
}

#[derive(Default)]
struct LinearPredictor {
    offset: f64,
    slope: f64,
}

impl Predictor for LinearPredictor {
    fn fit(mut self, x: impl Iterator<Item = f64>, y: impl Iterator<Item = f64>) -> Self {
        let mut n = 0.0;
        let mut sum_x = 0.0;
        let mut sum_y = 0.0;
        let mut sum_xx = 0.0;
        let mut sum_xy = 0.0;
        for (xi, yi) in x.zip(y) {
            n += 1.0;
            sum_x += xi;
            sum_y += yi;
            sum_xx += xi * xi;
            sum_xy += xi * yi;
        }
        self.slope = (sum_xy * n - sum_x * sum_y) / (sum_xx * n - sum_x * sum_x);
        self.offset = (sum_y - self.slope * sum_x) / n;
        self
    }

    fn predict<'a>(&self, x: impl Iterator<Item = f64> + 'a) -> Box<dyn Iterator<Item = f64> + 'a> {
        let slope = self.slope;
        let offset = self.offset;
        Box::new(x.map(move |xi| offset + slope * xi))
    }
}

struct LogYDecorator<P: Predictor> {
    decorated_predictor: P,
}

impl<P: Predictor> LogYDecorator<P> {
    pub fn new(decorated_predictor: P) -> Self {
        LogYDecorator {
            decorated_predictor,
        }
    }
}

impl<P: Predictor> Predictor for LogYDecorator<P> {
    fn fit(mut self, x: impl Iterator<Item = f64>, y: impl Iterator<Item = f64>) -> Self {
        let logy = y.map(f64::ln);
        self.decorated_predictor = self.decorated_predictor.fit(x, logy);
        self
    }

    fn predict<'a>(&self, x: impl Iterator<Item = f64> + 'a) -> Box<dyn Iterator<Item = f64> + 'a> {
        Box::new(self.decorated_predictor.predict(x).map(f64::exp))
    }
}

struct LogXDecorator<P: Predictor> {
    decorated_predictor: P,
}

impl<P: Predictor> LogXDecorator<P> {
    pub fn new(decorated_predictor: P) -> Self {
        LogXDecorator {
            decorated_predictor,
        }
    }
}

impl<P: Predictor> Predictor for LogXDecorator<P> {
    fn fit(mut self, x: impl Iterator<Item = f64>, y: impl Iterator<Item = f64>) -> Self {
        let logx = x.map(f64::ln);
        self.decorated_predictor = self.decorated_predictor.fit(logx, y);
        self
    }

    fn predict<'a>(&self, x: impl Iterator<Item = f64> + 'a) -> Box<dyn Iterator<Item = f64> + 'a> {
        let logx = x.map(f64::ln);
        Box::new(self.decorated_predictor.predict(logx))
    }
}

#[test]
fn predict() {
    let x_train = [2.0, 4.0, 6.0, 8.0];
    let y_train = [1.0, 2.0, 3.0, 4.0];

    let linear_predictor =
        LinearPredictor::default().fit(x_train.iter().copied(), y_train.iter().copied());
    println!(
        "linear predictor: {:?}",
        linear_predictor
            .predict(x_train.iter().copied())
            .collect::<Vec<_>>()
    );

    let logy_predictor = LogYDecorator::new(LinearPredictor::default())
        .fit(x_train.iter().copied(), y_train.iter().copied());
    println!(
        "log-y predictor: {:?}",
        logy_predictor
            .predict(x_train.iter().copied())
            .collect::<Vec<_>>()
    );

    let logx_predictor = LogXDecorator::new(LinearPredictor::default())
        .fit(x_train.iter().copied(), y_train.iter().copied());
    println!(
        "log-x predictor: {:?}",
        logx_predictor
            .predict(x_train.iter().copied())
            .collect::<Vec<_>>()
    );

    let loglog_predictor = LogXDecorator::new(LogYDecorator::new(LinearPredictor::default()))
        .fit(x_train.iter().copied(), y_train.iter().copied());
    println!(
        "log-log predictor: {:?}",
        loglog_predictor
            .predict(x_train.iter().copied())
            .collect::<Vec<_>>()
    );
}
